<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dimension reduction and clustering with singlecellutils • singlecellutils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Dimension reduction and clustering with singlecellutils">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singlecellutils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/singlecellutils-dimensionreduction.html">Dimension reduction and clustering with singlecellutils</a>
    </li>
    <li>
      <a href="../articles/singlecellutils-filtering.html">Filtering samples and features with singlecellutils</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jenzopr/singlecellutils">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Dimension reduction and clustering with singlecellutils</h1>
                        <h4 class="author">Jens Preussner</h4>
            
            <h4 class="date">2019-02-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jenzopr/singlecellutils/blob/master/vignettes/singlecellutils-dimensionreduction.Rmd"><code>vignettes/singlecellutils-dimensionreduction.Rmd</code></a></small>
      <div class="hidden name"><code>singlecellutils-dimensionreduction.Rmd</code></div>

    </div>

    
    
<p>This vignette will teach you how to use methods from <code>singlecellutils</code> to perform dimension reduction and clustering of samples. Methods from <code>singlecellutils</code> are enabled for <em>piping</em> (using <em><a href="https://CRAN.R-project.org/package=magrittr">magrittr</a></em>s pipe operator) and allow you to write code that is easily understandable.</p>
<div id="load-an-example-data-set---1k-brain-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#load-an-example-data-set---1k-brain-cells" class="anchor"></a>Load an example data set - 1k Brain cells</h2>
<p>Throughout this vignette, you’ll use the 1k Brain Cells dataset from <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_1k_v3">10xgenomics.com</a>. It ships with <code>singlecellutils</code> and is already preprocessed as a <code>SingleCellExperiment</code> data object from the <em><a href="https://bioconductor.org/packages/3.8/SingleCellExperiment">SingleCellExperiment</a></em> package. Here, we’ll load the object and perform filtering as described in the <a href="singlecellutils-filtering.html">Filtering samples and features with singlecellutils</a> vignette:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"neuron_1k_v3"</span>, <span class="dt">package =</span> <span class="st">"singlecellutils"</span>)</a>
<a class="sourceLine" id="cb1-2" title="2"></a>
<a class="sourceLine" id="cb1-3" title="3">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_samples</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"genefilter::kOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">k =</span> <span class="dv">1500</span>, <span class="dt">A =</span> <span class="dv">1</span>),</a>
<a class="sourceLine" id="cb1-5" title="5">                                       <span class="st">"singlecellutils::EpOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">A =</span> <span class="dv">3100</span>)),</a>
<a class="sourceLine" id="cb1-6" title="6">                                  <span class="dt">exprs_values =</span> <span class="st">"umi"</span>,</a>
<a class="sourceLine" id="cb1-7" title="7">                                  <span class="dt">tolerate =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_features</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"singlecellutils::MeanOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">A =</span> <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb1-9" title="9">                                        <span class="st">"genefilter::kOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">A =</span> <span class="dv">1</span>)),</a>
<a class="sourceLine" id="cb1-10" title="10">                                   <span class="dt">exprs_values =</span> <span class="st">"umi"</span>,</a>
<a class="sourceLine" id="cb1-11" title="11">                                   <span class="dt">tolerate =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-12" title="12"><span class="st">  </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/calculateQCMetrics">calculateQCMetrics</a></span>(<span class="dt">exprs_values =</span> <span class="st">"umi"</span>)</a></code></pre></div>
</div>
<div id="normalize-the-expression-using-scran-and-scater" class="section level2">
<h2 class="hasAnchor">
<a href="#normalize-the-expression-using-scran-and-scater" class="anchor"></a>Normalize the expression using scran and scater</h2>
<p>A typical step after filtering is normalization. Here, we’ll rely on a quick and simple calculation of cell-specific size factors using the <code>simpleSumFactors</code> function from the <em><a href="https://bioconductor.org/packages/3.8/scran">scran</a></em> package. The function adds the size factors to the <code>sizeFactors</code>-slot in <code>neuron_1k_v3</code>, which is picked up by <em><a href="https://bioconductor.org/packages/3.8/scater">scater</a></em>’s <code>normalize</code> function:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="st">  </span>scran<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scran/topics/simpleSumFactors">simpleSumFactors</a></span>(<span class="dt">assay.type =</span> <span class="st">"umi"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="st">  </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(<span class="dt">exprs_values =</span> <span class="st">"umi"</span>)</a></code></pre></div>
</div>
<div id="performing-dimension-reduction-using-umap" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-dimension-reduction-using-umap" class="anchor"></a>Performing dimension reduction using UMAP</h2>
<blockquote>
<p>UMAP (Uniform Manifold Approximation and Projection) is a novel manifold learning technique for dimension reduction. […] The UMAP algorithm is competitive with t-SNE for visualization quality, and arguably preserves more of the global structure with superior run time performance. Furthermore, UMAP has no computational restrictions on embedding dimension, making it viable as a general purpose dimension reduction technique for machine learning.</p>
<p><em>from the <a href="https://arxiv.org/abs/1802.03426">UMAP paper</a> abstract</em></p>
</blockquote>
<p>Two parameters determine UMAPs performance:</p>
<p>The <strong>number of neighboring points</strong> determines how much of the global structure is being preserved at the loss of local structure. Larger values preserve more of the global structure and values in the range of 5 to 50 are mostly fine, with 10-15 being sensible defaults.</p>
<p>The <strong>minimum distance</strong> controls how tightly the embedding is allowed to compress points together. Larger values ensure embedded points are more evenly distributed, while smaller values allow the algorithm to optimise more accurately with regard to local structure. Sensible values are in the range 0.001 to 0.5, with 0.1 being a reasonable default.</p>
<p>We’ll first explore, how different values of one of those two parameters change the embedding, when keeping the other value fixed:</p>
<div class="figure">
<img src="singlecellutils-dimensionreduction_files/figure-html/umap-explore-neighbors-1.png" alt="Performance of UMAP across differen values of `n_neighbors`." width="768"><p class="caption">
Performance of UMAP across differen values of <code>n_neighbors</code>.
</p>
</div>
<div class="figure">
<img src="singlecellutils-dimensionreduction_files/figure-html/umap-explore-distance-1.png" alt="Performance of UMAP across differen values of `min_dist`." width="768"><p class="caption">
Performance of UMAP across differen values of <code>min_dist</code>.
</p>
</div>
<p>As we can see from above, a value of <code>20</code> for <code>n_neighbors</code> should be chosen at minimum, whereas the <code>min_dist</code> should not exceed <code>0.1</code> to permit reliable identification of subclusters.</p>
<p>We can now use UMAP and visualize the resulting dimension reduction in several ways:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/reduce_dimension">reduce_dimension</a></span>(<span class="dt">flavor =</span> <span class="st">"umap"</span>,</a>
<a class="sourceLine" id="cb3-3" title="3">                                    <span class="dt">exprs_values =</span> <span class="st">"logcounts"</span>,</a>
<a class="sourceLine" id="cb3-4" title="4">                                    <span class="dt">n_neighbors =</span> 20L,</a>
<a class="sourceLine" id="cb3-5" title="5">                                    <span class="dt">min_dist =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb3-6" title="6">                                    <span class="dt">seed =</span> <span class="dv">2304</span>)</a></code></pre></div>
<p>and visualize the dimension reduction with different quality criteria (see <a href="#vis">Section visualization</a> for source code examples):</p>
<div class="figure">
<img src="singlecellutils-dimensionreduction_files/figure-html/umap-figure-1.png" alt="UMAP dimension reduction on the `neuron_1k_v3` dataset, using `n_neighbors = 20` and `min_dist = 0.1` as input parameters." width="768"><p class="caption">
UMAP dimension reduction on the <code>neuron_1k_v3</code> dataset, using <code>n_neighbors = 20</code> and <code>min_dist = 0.1</code> as input parameters.
</p>
</div>
</div>
<div id="clustering-samples-using-hdbscan" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-samples-using-hdbscan" class="anchor"></a>Clustering samples using HDBSCAN</h2>
<p><strong>H</strong>ierarchical <strong>D</strong>ensity-<strong>B</strong>ased <strong>S</strong>patial <strong>C</strong>lustering of <strong>A</strong>pplications with <strong>N</strong>oise (HDBSCAN)<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, is a new density-based clustering algorithm that extends DBSCAN<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> by converting it into a hierarchical clustering algorithm, and then using a technique to extract a flat clustering based in the stability of clusters. This allows HDBSCAN to find clusters of varying densities (unlike DBSCAN), and be more robust to parameters.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/add_clustering">add_clustering</a></span>(<span class="dt">flavor =</span> <span class="st">"hdbscan"</span>,</a>
<a class="sourceLine" id="cb4-3" title="3">                                  <span class="dt">use_dimred =</span> <span class="st">"umap"</span>,</a>
<a class="sourceLine" id="cb4-4" title="4">                                  <span class="dt">column =</span> <span class="st">".clustering_911"</span>,</a>
<a class="sourceLine" id="cb4-5" title="5">                                  <span class="dt">min_samples =</span> 9L,</a>
<a class="sourceLine" id="cb4-6" title="6">                                  <span class="dt">min_cluster_size =</span> 11L)</a></code></pre></div>
<div id="vis" class="section level3">
<h3 class="hasAnchor">
<a href="#vis" class="anchor"></a>Visualization of dimension reduction and clustering</h3>
<p>Visualization of different clusterings (e.g. different parameters) can be performed with functions from the <em><a href="https://bioconductor.org/packages/3.8/scater">scater</a></em> package:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"></a>
<a class="sourceLine" id="cb5-2" title="2">hdbscan_p1 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(neuron_1k_v3, <span class="dt">use_dimred =</span> <span class="st">"umap"</span>, <span class="dt">colour_by =</span> <span class="st">".clustering_911"</span>, <span class="dt">add_ticks =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-3" title="3">hdbscan_p2 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(neuron_1k_v3, <span class="dt">use_dimred =</span> <span class="st">"umap"</span>, <span class="dt">colour_by =</span> <span class="st">".clustering_1719"</span>, <span class="dt">add_ticks =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-4" title="4">hdbscan_p3 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(neuron_1k_v3, <span class="dt">use_dimred =</span> <span class="st">"umap"</span>, <span class="dt">colour_by =</span> <span class="st">".clustering_2123"</span>, <span class="dt">add_ticks =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-5" title="5">hdbscan_p4 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(neuron_1k_v3, <span class="dt">use_dimred =</span> <span class="st">"umap"</span>, <span class="dt">colour_by =</span> <span class="st">".clustering_2325"</span>, <span class="dt">add_ticks =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-6" title="6"></a>
<a class="sourceLine" id="cb5-7" title="7">gridExtra<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/gridExtra/topics/arrangeGrob">grid.arrange</a></span>(hdbscan_p1, hdbscan_p2, hdbscan_p3, hdbscan_p4, <span class="dt">nrow =</span> <span class="dv">2</span>)</a></code></pre></div>
<div class="figure">
<img src="singlecellutils-dimensionreduction_files/figure-html/hdbscan-visualization-1.png" alt="UMAP dimension reduction on the `neuron_1k_v3` dataset, coloured by different HDBSCAN clusterings." width="768"><p class="caption">
UMAP dimension reduction on the <code>neuron_1k_v3</code> dataset, coloured by different HDBSCAN clusterings.
</p>
</div>
</div>
</div>
<div id="using-pipes-to-reduce-dimension-and-perform-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#using-pipes-to-reduce-dimension-and-perform-clustering" class="anchor"></a>Using pipes to reduce dimension and perform clustering</h2>
<p>The code snippets from above for normalization, dimension reduction and clustering can easily be combined into one <em>pipeline</em>. This will reduce the amount of code you’ll have to write, enhances visibility of what functions are applied to the neuron_1k_v3 object and normally speeds up your analysis as well. Here we go:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="st">  </span>scran<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scran/topics/simpleSumFactors">simpleSumFactors</a></span>(<span class="dt">assay.type =</span> <span class="st">"umi"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-3" title="3"><span class="st">  </span>scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(<span class="dt">exprs_values =</span> <span class="st">"umi"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/reduce_dimension">reduce_dimension</a></span>(<span class="dt">flavor =</span> <span class="st">"umap"</span>,</a>
<a class="sourceLine" id="cb6-5" title="5">                                    <span class="dt">exprs_values =</span> <span class="st">"logcounts"</span>,</a>
<a class="sourceLine" id="cb6-6" title="6">                                    <span class="dt">n_neighbors =</span> 20L,</a>
<a class="sourceLine" id="cb6-7" title="7">                                    <span class="dt">min_dist =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb6-8" title="8">                                    <span class="dt">seed =</span> <span class="dv">2304</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/add_clustering">add_clustering</a></span>(<span class="dt">flavor =</span> <span class="st">"hdbscan"</span>,</a>
<a class="sourceLine" id="cb6-10" title="10">                                  <span class="dt">use_dimred =</span> <span class="st">"umap"</span>,</a>
<a class="sourceLine" id="cb6-11" title="11">                                  <span class="dt">min_samples =</span> 21L,</a>
<a class="sourceLine" id="cb6-12" title="12">                                  <span class="dt">min_cluster_size =</span> 23L)</a></code></pre></div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>see McInnes et. al., hdbscan: Hierarchical density based clustering (2017)<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>see Ester et. al., A density-based algorithm for discovering clusters in large spatial databases with noise (1996)<a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-an-example-data-set---1k-brain-cells">Load an example data set - 1k Brain cells</a></li>
      <li><a href="#normalize-the-expression-using-scran-and-scater">Normalize the expression using scran and scater</a></li>
      <li><a href="#performing-dimension-reduction-using-umap">Performing dimension reduction using UMAP</a></li>
      <li><a href="#clustering-samples-using-hdbscan">Clustering samples using HDBSCAN</a></li>
      <li><a href="#using-pipes-to-reduce-dimension-and-perform-clustering">Using pipes to reduce dimension and perform clustering</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jens Preussner.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
