<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Filtering samples and features with singlecellutils • singlecellutils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Filtering samples and features with singlecellutils">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singlecellutils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/singlecellutils-dimensionreduction.html">Dimension reduction and clustering with singlecellutils</a>
    </li>
    <li>
      <a href="../articles/singlecellutils-filtering.html">Filtering samples and features with singlecellutils</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jenzopr/singlecellutils">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Filtering samples and features with singlecellutils</h1>
                        <h4 class="author">Jens Preussner</h4>
            
            <h4 class="date">2019-02-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jenzopr/singlecellutils/blob/master/vignettes/singlecellutils-filtering.Rmd"><code>vignettes/singlecellutils-filtering.Rmd</code></a></small>
      <div class="hidden name"><code>singlecellutils-filtering.Rmd</code></div>

    </div>

    
    
<p>This vignette will teach you how to use methods from <code>singlecellutils</code> to detect and filter low quality samples and exclude features that are lowly expressed. Methods from <code>singlecellutils</code> are enabled for <em>piping</em> (using <em><a href="https://CRAN.R-project.org/package=magrittr">magrittr</a></em>s pipe operator) and allow you to write code that is easily understandable.</p>
<div id="load-an-example-data-set---1k-brain-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#load-an-example-data-set---1k-brain-cells" class="anchor"></a>Load an example data set - 1k Brain cells</h2>
<p>Throughout this vignette, you’ll use the 1k Brain Cells dataset from <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_1k_v3">10xgenomics.com</a>. It ships with <code>singlecellutils</code> and is already preprocessed as a <code>SingleCellExperiment</code> data object from the <em><a href="https://bioconductor.org/packages/3.8/SingleCellExperiment">SingleCellExperiment</a></em> package. Just load the object like below and you’re good to go:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"neuron_1k_v3"</span>, <span class="dt">package =</span> <span class="st">"singlecellutils"</span>)</a></code></pre></div>
</div>
<div id="low-quality-sample-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#low-quality-sample-filtering" class="anchor"></a>Low quality sample filtering</h2>
<p>Samples of low quality are typically associated with the detection of only a few features (e.g. <em>genes</em>) or have generated only a low number of sequencing reads due to deteriorated DNA. During preprocessing of dataset loaded above, we also ran the <code>calculateQCMetrics()</code> function from the <em><a href="https://bioconductor.org/packages/3.8/scater">scater</a></em> package, that adds such important quality criteria to the <code>colData</code> slot of the <code>neuron_1k_v3</code> object. Take a quick look to get a feeling for the information that is available:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(neuron_1k_v3))</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co">#&gt; DataFrame with 6 rows and 9 columns</span></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="co">#&gt;                    is_cell_control total_features_by_umi</span></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co">#&gt;                          &lt;logical&gt;             &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">#&gt; AAACGAATCAAAGCCT-1           FALSE                   417</span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">#&gt; AAACGCTGTAATGTGA-1           FALSE                  2744</span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">#&gt; AAACGCTGTCCTGGGT-1           FALSE                  3603</span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co">#&gt; AAAGAACCAGGACATG-1           FALSE                  2386</span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="co">#&gt; AAAGGTACACACGGTC-1           FALSE                  4748</span></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="co">#&gt; AAAGTCCAGTCACTAC-1           FALSE                  5104</span></a>
<a class="sourceLine" id="cb2-11" title="11"><span class="co">#&gt;                    log10_total_features_by_umi total_umi  log10_total_umi</span></a>
<a class="sourceLine" id="cb2-12" title="12"><span class="co">#&gt;                                      &lt;numeric&gt; &lt;numeric&gt;        &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb2-13" title="13"><span class="co">#&gt; AAACGAATCAAAGCCT-1            2.62117628177504      1023 3.01029995663981</span></a>
<a class="sourceLine" id="cb2-14" title="14"><span class="co">#&gt; AAACGCTGTAATGTGA-1            3.43854234878611      7210 3.85799549556092</span></a>
<a class="sourceLine" id="cb2-15" title="15"><span class="co">#&gt; AAACGCTGTCCTGGGT-1            3.55678478230703      9883 3.99493273673004</span></a>
<a class="sourceLine" id="cb2-16" title="16"><span class="co">#&gt; AAAGAACCAGGACATG-1            3.37785241900675      5290 3.72353776153206</span></a>
<a class="sourceLine" id="cb2-17" title="17"><span class="co">#&gt; AAAGGTACACACGGTC-1            3.67660216958202     18676 4.27130711878496</span></a>
<a class="sourceLine" id="cb2-18" title="18"><span class="co">#&gt; AAAGTCCAGTCACTAC-1            3.70799574642293     20196 4.30528686547613</span></a>
<a class="sourceLine" id="cb2-19" title="19"><span class="co">#&gt;                    pct_umi_in_top_50_features pct_umi_in_top_100_features</span></a>
<a class="sourceLine" id="cb2-20" title="20"><span class="co">#&gt;                                     &lt;numeric&gt;                   &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb2-21" title="21"><span class="co">#&gt; AAACGAATCAAAGCCT-1           64.1251221896383            69.0127077223851</span></a>
<a class="sourceLine" id="cb2-22" title="22"><span class="co">#&gt; AAACGCTGTAATGTGA-1           29.9583911234397            38.4466019417476</span></a>
<a class="sourceLine" id="cb2-23" title="23"><span class="co">#&gt; AAACGCTGTCCTGGGT-1           22.2908023879389            29.8391176768188</span></a>
<a class="sourceLine" id="cb2-24" title="24"><span class="co">#&gt; AAAGAACCAGGACATG-1           22.5141776937618            30.4914933837429</span></a>
<a class="sourceLine" id="cb2-25" title="25"><span class="co">#&gt; AAAGGTACACACGGTC-1           22.4780466909402            34.2150353394731</span></a>
<a class="sourceLine" id="cb2-26" title="26"><span class="co">#&gt; AAAGTCCAGTCACTAC-1           23.9255298078827            32.9570211923153</span></a>
<a class="sourceLine" id="cb2-27" title="27"><span class="co">#&gt;                    pct_umi_in_top_200_features pct_umi_in_top_500_features</span></a>
<a class="sourceLine" id="cb2-28" title="28"><span class="co">#&gt;                                      &lt;numeric&gt;                   &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb2-29" title="29"><span class="co">#&gt; AAACGAATCAAAGCCT-1            78.7878787878788                         100</span></a>
<a class="sourceLine" id="cb2-30" title="30"><span class="co">#&gt; AAACGCTGTAATGTGA-1            47.3509015256588            61.5811373092926</span></a>
<a class="sourceLine" id="cb2-31" title="31"><span class="co">#&gt; AAACGCTGTCCTGGGT-1            39.2492158251543            53.8298087625215</span></a>
<a class="sourceLine" id="cb2-32" title="32"><span class="co">#&gt; AAAGAACCAGGACATG-1            40.1701323251418            56.9754253308129</span></a>
<a class="sourceLine" id="cb2-33" title="33"><span class="co">#&gt; AAAGGTACACACGGTC-1            44.8704219318912            58.2512315270936</span></a>
<a class="sourceLine" id="cb2-34" title="34"><span class="co">#&gt; AAAGTCCAGTCACTAC-1            43.1867696573579             56.387403446227</span></a></code></pre></div>
<p>You’re particularly interested in the <code>total_features_by_umi</code> column, whose values count the number of detected features, and in the <code>total_umi</code> column, whose values count the number of <em>unique molecular identifiers</em> (UMI) associated with each sample.</p>
<p>Let’s check how those two quality criteria distribute, by plotting a violin plot, where each sample is represented by a dot:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotColData">plotColData</a></span>(neuron_1k_v3, <span class="dt">y =</span> <span class="st">"total_features_by_umi"</span>)</a>
<a class="sourceLine" id="cb3-2" title="2">scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotColData">plotColData</a></span>(neuron_1k_v3, <span class="dt">y =</span> <span class="st">"log10_total_umi"</span>)</a></code></pre></div>
<div class="figure">
<img src="singlecellutils-filtering_files/figure-html/sample-filtering-distribution-eval-1.png" alt="Distribution of `total_features_by_umi` and `total_umi` (log-scaled) across all samples in the `neuron_1k_v3` dataset." width="700"><img src="singlecellutils-filtering_files/figure-html/sample-filtering-distribution-eval-2.png" alt="Distribution of `total_features_by_umi` and `total_umi` (log-scaled) across all samples in the `neuron_1k_v3` dataset." width="700"><p class="caption">
Distribution of <code>total_features_by_umi</code> and <code>total_umi</code> (log-scaled) across all samples in the <code>neuron_1k_v3</code> dataset.
</p>
</div>
<p>From the figure above, you can see that roughly 1500 features are detected in the vast majority of samples. Further, most of the cells are associated with <span class="math inline">\(10^{3.5}\)</span> (thats slightly more than 3100) or more distinct UMIs. Samples below those values (or left of the red vertical lines) are potentially outliers and could be removed before downstream analysis<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<blockquote>
<p>The <code>singlecellutils</code> package adopts the approach of the <em><a href="https://bioconductor.org/packages/3.8/genefilter">genefilter</a></em> package to filter samples and features. The idea is to combine output from multiple <em>filter functions</em>, each returning <code>TRUE</code> or <code>FALSE</code> for every sample/feature, indicating if the sample/feature should be filtered out or not. However, <code>singlecellutils</code> extends the functionality of <em><a href="https://bioconductor.org/packages/3.8/genefilter">genefilter</a></em> to allow <em>OR</em> conjunctions between filters as well.</p>
</blockquote>
<p>Create two sample filters, using the <code>kOverA</code> function from <code>genefilter</code> and the <code>EpOverA</code> function from <code>singlecellutils</code>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">sample_filter &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"genefilter::kOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">k =</span> <span class="dv">1500</span>, <span class="dt">A =</span> <span class="dv">1</span>),</a>
<a class="sourceLine" id="cb4-2" title="2">                      <span class="st">"singlecellutils::EpOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">A =</span> <span class="dv">3100</span>))</a></code></pre></div>
<p><strong>Explanation</strong>: <code><a href="https://www.rdocumentation.org/packages/genefilter/topics/kOverA">genefilter::kOverA</a></code> is a filter function that returns <code>TRUE</code>, if <code>k</code> inspected values are above a value of <code>A</code>. In our case, <code>k = 1500</code> features have to exceed a UMI count of <code>A = 1</code>, indicating expression. <code><a href="../reference/filterfunctions.html">singlecellutils::EpOverA</a></code> is a filter function, that returns <code>TRUE</code> if the aggregate of values is above <code>A</code>. In our case, samples have to exceed an aggregated UMI count of <code>A = 3100</code>.</p>
<p>For example, after a call to</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb5-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_samples</a></span>(sample_filter, <span class="dt">exprs_values =</span> <span class="st">"umi"</span>)</a></code></pre></div>
<p>the <code>neuron_1k_v3</code> object contains 601 remaining samples. If you would like to combine filters with an <em>OR</em> conjunction use the <code>tolerate</code> argument of <code>filter_samples</code>:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_samples</a></span>(sample_filter, <span class="dt">exprs_values =</span> <span class="st">"umi"</span>, <span class="dt">tolerate =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>In this case, 1108 high quality samples remain.</p>
</div>
<div id="filtering-lowly-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-lowly-expressed-genes" class="anchor"></a>Filtering lowly expressed genes</h2>
<p>Similar to low quality samples, lowly expressed features can also obfuscate downstream analysis (e.g. <em>dimension reduction</em> or <em>clustering</em>) and should be removed as well. Again, take a quick look at the output of the <code>calculateQCMetrics()</code> function, this time for <em>feature-based</em> quality measures:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">rowData</a></span>(neuron_1k_v3))</a>
<a class="sourceLine" id="cb7-2" title="2"><span class="co">#&gt; DataFrame with 6 rows and 10 columns</span></a>
<a class="sourceLine" id="cb7-3" title="3"><span class="co">#&gt;                               gene_id   gene_name         feature</span></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt;</span></a>
<a class="sourceLine" id="cb7-5" title="5"><span class="co">#&gt; ENSMUSG00000051951 ENSMUSG00000051951        Xkr4 Gene Expression</span></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="co">#&gt; ENSMUSG00000089699 ENSMUSG00000089699      Gm1992 Gene Expression</span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="co">#&gt; ENSMUSG00000102343 ENSMUSG00000102343     Gm37381 Gene Expression</span></a>
<a class="sourceLine" id="cb7-8" title="8"><span class="co">#&gt; ENSMUSG00000025900 ENSMUSG00000025900         Rp1 Gene Expression</span></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="co">#&gt; ENSMUSG00000025902 ENSMUSG00000025902       Sox17 Gene Expression</span></a>
<a class="sourceLine" id="cb7-10" title="10"><span class="co">#&gt; ENSMUSG00000104328 ENSMUSG00000104328     Gm37323 Gene Expression</span></a>
<a class="sourceLine" id="cb7-11" title="11"><span class="co">#&gt;                    is_feature_control             mean_umi</span></a>
<a class="sourceLine" id="cb7-12" title="12"><span class="co">#&gt;                             &lt;logical&gt;            &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb7-13" title="13"><span class="co">#&gt; ENSMUSG00000051951              FALSE    0.116064565718678</span></a>
<a class="sourceLine" id="cb7-14" title="14"><span class="co">#&gt; ENSMUSG00000089699              FALSE 0.000768639508070715</span></a>
<a class="sourceLine" id="cb7-15" title="15"><span class="co">#&gt; ENSMUSG00000102343              FALSE                    0</span></a>
<a class="sourceLine" id="cb7-16" title="16"><span class="co">#&gt; ENSMUSG00000025900              FALSE  0.00153727901614143</span></a>
<a class="sourceLine" id="cb7-17" title="17"><span class="co">#&gt; ENSMUSG00000025902              FALSE   0.0653343581860108</span></a>
<a class="sourceLine" id="cb7-18" title="18"><span class="co">#&gt; ENSMUSG00000104328              FALSE 0.000768639508070715</span></a>
<a class="sourceLine" id="cb7-19" title="19"><span class="co">#&gt;                          log10_mean_umi n_cells_by_umi pct_dropout_by_umi</span></a>
<a class="sourceLine" id="cb7-20" title="20"><span class="co">#&gt;                               &lt;numeric&gt;      &lt;integer&gt;          &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb7-21" title="21"><span class="co">#&gt; ENSMUSG00000051951   0.0476893198024886            120   90.7763259031514</span></a>
<a class="sourceLine" id="cb7-22" title="22"><span class="co">#&gt; ENSMUSG00000089699 0.000333687670586935              1   99.9231360491929</span></a>
<a class="sourceLine" id="cb7-23" title="23"><span class="co">#&gt; ENSMUSG00000102343                    0              0                100</span></a>
<a class="sourceLine" id="cb7-24" title="24"><span class="co">#&gt; ENSMUSG00000025900 0.000667119150998434              2   99.8462720983859</span></a>
<a class="sourceLine" id="cb7-25" title="25"><span class="co">#&gt; ENSMUSG00000025902   0.0274859337142017             11   99.1544965411222</span></a>
<a class="sourceLine" id="cb7-26" title="26"><span class="co">#&gt; ENSMUSG00000104328 0.000333687670586935              1   99.9231360491929</span></a>
<a class="sourceLine" id="cb7-27" title="27"><span class="co">#&gt;                    total_umi   log10_total_umi</span></a>
<a class="sourceLine" id="cb7-28" title="28"><span class="co">#&gt;                    &lt;numeric&gt;         &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb7-29" title="29"><span class="co">#&gt; ENSMUSG00000051951       151  2.18184358794477</span></a>
<a class="sourceLine" id="cb7-30" title="30"><span class="co">#&gt; ENSMUSG00000089699         1 0.301029995663981</span></a>
<a class="sourceLine" id="cb7-31" title="31"><span class="co">#&gt; ENSMUSG00000102343         0                 0</span></a>
<a class="sourceLine" id="cb7-32" title="32"><span class="co">#&gt; ENSMUSG00000025900         2 0.477121254719662</span></a>
<a class="sourceLine" id="cb7-33" title="33"><span class="co">#&gt; ENSMUSG00000025902        85  1.93449845124357</span></a>
<a class="sourceLine" id="cb7-34" title="34"><span class="co">#&gt; ENSMUSG00000104328         1 0.301029995663981</span></a></code></pre></div>
<p>Maybe you’re particularly interested in the <code>mean_umi</code> column, whose values represent the <em>mean</em> UMI count of the feature across all samples, and in the <code>n_cells_by_umi</code> column, whose values count the number of samples that express the feature.</p>
<p>Let’s check how those two quality criteria distribute, by plotting a violin plot, where each feature is represented by a dot:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotRowData">plotRowData</a></span>(neuron_1k_v3, <span class="dt">y =</span> <span class="st">"mean_umi"</span>)</a>
<a class="sourceLine" id="cb8-2" title="2">scater<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotRowData">plotRowData</a></span>(neuron_1k_v3, <span class="dt">y =</span> <span class="st">"n_cells_by_umi"</span>)</a></code></pre></div>
<div class="figure">
<img src="singlecellutils-filtering_files/figure-html/feature-filtering-distribution-eval-1.png" alt="Distribution of `mean_umi` and `n_cells_by_umi` across all features in the `neuron_1k_v3` dataset." width="700"><img src="singlecellutils-filtering_files/figure-html/feature-filtering-distribution-eval-2.png" alt="Distribution of `mean_umi` and `n_cells_by_umi` across all features in the `neuron_1k_v3` dataset." width="700"><p class="caption">
Distribution of <code>mean_umi</code> and <code>n_cells_by_umi</code> across all features in the <code>neuron_1k_v3</code> dataset.
</p>
</div>
<p>From the figure above, you can see that most features are expressed at very low levels (the red bar indicates a mean of 0.1) and only in a fraction of samples (the red bar indicates 20 samples). We will demonstrate feature filtering using two <em>filter functions</em> from <code>singlecellutils</code> and <code>genefilter</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">feature_filter &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"singlecellutils::MeanOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">A =</span> <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb9-2" title="2">                      <span class="st">"genefilter::kOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">A =</span> <span class="dv">1</span>))</a></code></pre></div>
<p><strong>Explanation</strong>: <code><a href="../reference/filterfunctions.html">singlecellutils::MeanOverA</a></code> is a filter function that returns <code>TRUE</code>, if the mean of the inspected values are above a value of <code>A</code>. In our case, the mean UMI count of a feature has to be above <code>A = 0.1</code> for the feature to remain in the dataset. <code><a href="https://www.rdocumentation.org/packages/genefilter/topics/kOverA">genefilter::kOverA</a></code> is a filter function that returns <code>TRUE</code>, if <code>k</code> inspected values are above a value of <code>A</code>. In our case, features have to be expressed in at least <code>k = 10</code> samples.</p>
<p>For example, after a call to</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_features</a></span>(feature_filter, <span class="dt">exprs_values =</span> <span class="st">"umi"</span>)</a></code></pre></div>
<p>the <code>neuron_1k_v3</code> object contains 9237 remaining features. If you would like to combine filters with an <em>OR</em> conjunction use the <code>tolerate</code> argument of <code>filter_features</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_features</a></span>(feature_filter, <span class="dt">exprs_values =</span> <span class="st">"umi"</span>, <span class="dt">tolerate =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>In this case, 10173 features remain in the dataset.</p>
</div>
<div id="using-pipes-to-filter-samples-and-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#using-pipes-to-filter-samples-and-genes" class="anchor"></a>Using pipes to filter samples and genes</h2>
<p>The examples showing sample and feature filtering from above can easily be combined into one <em>pipeline</em>. This will reduce the amount of code you’ll have to write, enhances visibility of what functions are applied to the <code>neuron_1k_v3</code> object and normally speeds up your analysis as well. Here we go:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">neuron_1k_v3 <span class="op">%&lt;&gt;%</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_samples</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"genefilter::kOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">k =</span> <span class="dv">1500</span>, <span class="dt">A =</span> <span class="dv">1</span>),</a>
<a class="sourceLine" id="cb12-3" title="3">                                       <span class="st">"singlecellutils::EpOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">A =</span> <span class="dv">3100</span>)),</a>
<a class="sourceLine" id="cb12-4" title="4">                                  <span class="dt">exprs_values =</span> <span class="st">"umi"</span>,</a>
<a class="sourceLine" id="cb12-5" title="5">                                  <span class="dt">tolerate =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-6" title="6"><span class="st">  </span>singlecellutils<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/singlecellutils/topics/filter">filter_features</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"singlecellutils::MeanOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">A =</span> <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb12-7" title="7">                                        <span class="st">"genefilter::kOverA"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">A =</span> <span class="dv">1</span>)),</a>
<a class="sourceLine" id="cb12-8" title="8">                                   <span class="dt">exprs_values =</span> <span class="st">"umi"</span>,</a>
<a class="sourceLine" id="cb12-9" title="9">                                   <span class="dt">tolerate =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb12-10" title="10">                                    </a></code></pre></div>
<p>Using only two lines, the final object is filtered for low quality samples and lowly expressed features and now contains data from 10234 features measured across 1108 samples. Happy downstream analysis ;-)</p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Feel free to see <em><a href="https://bioconductor.org/packages/3.8/scater">scater</a></em>s function <code>isOutlier</code> for a more sophisticated approach to detect outlier cells.<a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-an-example-data-set---1k-brain-cells">Load an example data set - 1k Brain cells</a></li>
      <li><a href="#low-quality-sample-filtering">Low quality sample filtering</a></li>
      <li><a href="#filtering-lowly-expressed-genes">Filtering lowly expressed genes</a></li>
      <li><a href="#using-pipes-to-filter-samples-and-genes">Using pipes to filter samples and genes</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jens Preussner.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
